### PR TITLE
MULE-18644: ValueProvider, Metadata fails due token is expired with OAuth Authorization Code Grant

### PR DESCRIPTION
**Pull request by ndinu**

### CODE REVIEW

**Comment by ndinu** [comment]
--validate

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java [hunk 1]
```diff
@@ -85,11 +87,11 @@
     try {
       final Map<Integer, ParameterModel> partsByOrder = getPartOrderMapping(keyParts);
       Set<MetadataKey> metadataKeys;
-      if (keyResolver instanceof PartialTypeKeysResolver && hasInitialLevel(partialKey, partsByOrder, reflectionCache)) {
-        metadataKeys = singleton(((PartialTypeKeysResolver) keyResolver).resolveChilds(context, partialKey));
-      } else {
-        metadataKeys = keyResolver.getKeys(context);
-      }
+
+      metadataKeys = context instanceof ConnectionProviderAwareMetadataContext
```
Comment (gabrieldalborgo) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java:91]: Unify with the definition of the variable
`Set<MetadataKey> metadataKeys = context instanceOf ConnectionProviderAwareMetadaContext...`

**Comment by ndinu** [comment]
--validate

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.mule.runtime.api.connection.ConnectionProvider;
+import org.mule.runtime.api.metadata.MetadataContext;
+
+import java.util.Optional;
+
+/**
+ * A {@link MetadataContext} which is aware of the {@link ConnectionProvider} used to provide its connection.
+ *
+ * @since 4.4.0, 4.3.1
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:17]: why 4.3.1?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ */
+public interface ConnectionProviderAwareMetadataContext extends MetadataContext {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:19]: why the specialization intstead of adding this to the general metadataContext ?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java [hunk 4]
```diff
@@ -84,12 +86,11 @@
...
     try {
       final Map<Integer, ParameterModel> partsByOrder = getPartOrderMapping(keyParts);
-      Set<MetadataKey> metadataKeys;
-      if (keyResolver instanceof PartialTypeKeysResolver && hasInitialLevel(partialKey, partsByOrder, reflectionCache)) {
-        metadataKeys = singleton(((PartialTypeKeysResolver) keyResolver).resolveChilds(context, partialKey));
-      } else {
-        metadataKeys = keyResolver.getKeys(context);
-      }
+      Set<MetadataKey> metadataKeys = context instanceof ConnectionProviderAwareMetadataContext
+          ? getWithTokenRefreshIfNecessary(((ConnectionProviderAwareMetadataContext) context).getConnectionProvider()
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java:90]: this ternary construct is all over the place. Generalize it more.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/ExtensionsOAuthUtils.java [hunk 5]
```diff
@@ -92,9 +94,12 @@ public static ClientCredentialsLocation toCredentialsLocation(CredentialsPlaceme
 
   public static OAuthConnectionProviderWrapper getOAuthConnectionProvider(ExecutionContextAdapter operationContext) {
     ConfigurationInstance config = ((ConfigurationInstance) operationContext.getConfiguration().get());
-    ConnectionProvider provider =
-        unwrapProviderWrapper(config.getConnectionProvider().get(), OAuthConnectionProviderWrapper.class);
-    return provider instanceof OAuthConnectionProviderWrapper ? (OAuthConnectionProviderWrapper) provider : null;
+    return getOAuthConnectionProvider(config.getConnectionProvider().get());
+  }
+
+  public static OAuthConnectionProviderWrapper getOAuthConnectionProvider(ConnectionProvider provider) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/ExtensionsOAuthUtils.java:100]: jdoc

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/ExtensionsOAuthUtils.java [hunk 6]
```diff
@@ -195,19 +200,57 @@ public static OAuthConnectionProviderWrapper getOAuthConnectionProvider(Executio
     }
   }
 
+  public static <T> T getWithTokenRefreshIfNecessary(ConnectionProvider connectionProvider, CheckedSupplier<T> supplier)
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/ExtensionsOAuthUtils.java:203]: add jdoc and rename to `withRefreshToken`

**Review by marianogonzalez — CHANGES_REQUESTED** [review]
you need the same thing for sample data

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java (hunk 2 above)
Comment (ndinu) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:17]: This should be cherry-picked to solve this issue for 4.3.x versions. If we decide that that is not necessary, we can add it only in 4.4.0.

Comment (marianogonzalez) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:17]: the bug cannot be reproduced without OCS, which will only be available with 4.4.0

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java (hunk 3 above)
Comment (ndinu) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:19]: MetadataContext is an api interface that developers use when resolving metadata. We don't want them to access to the instance of the connection provider. This ConnectionProviderAwareMetadataContext is an internal interface.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java (hunk 4 above)
Comment (ndinu) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeysDelegate.java:90]: Do you mean like having a util method that receives the context and a supplier and have this instanceof logic there?

**Comment by ndinu** [comment]
Regarding sample data, should I create a new jira issue to tackle the refresh there? @marianogonzalez

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java (hunk 2 above)
Comment (ndinu) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ConnectionProviderAwareMetadataContext.java:17]: 👍

**Comment by marianogonzalez** [comment]
Your call. It was just a reminder

**Review by marianogonzalez — APPROVED** [review]

**Pull request merged**